src/main/java/io/lettuce/core/protocol/DefaultEndpoint.java

```diff
@@ -379,38 +380,44 @@ private void writeToChannelAndFlush(RedisCommand<?, ?, ?> command) {
         if (reliability == Reliability.AT_MOST_ONCE) {
             // cancel on exceptions and remove from queue, because there is no housekeeping
             channelFuture.addListener(AtMostOnceWriteListener.newInstance(this, command));
-        }
-
-        if (reliability == Reliability.AT_LEAST_ONCE) {
+        } else if (reliability == Reliability.AT_LEAST_ONCE) {
             // commands are ok to stay within the queue, reconnect will retrigger them
             channelFuture.addListener(RetryListener.newInstance(this, command));
         }
     }
 
     private void writeToChannelAndFlush(Collection<? extends RedisCommand<?, ?, ?>> commands) {
+        final Channel chan = this.channel;
+        final EventLoop eventLoop = chan.eventLoop();
 
         QUEUE_SIZE.addAndGet(this, commands.size());
 
         if (reliability == Reliability.AT_MOST_ONCE) {
-
-            // cancel on exceptions and remove from queue, because there is no housekeeping
-            for (RedisCommand<?, ?, ?> command : commands) {
-                channelWrite(command).addListener(AtMostOnceWriteListener.newInstance(this, command));
-            }
+            executeInEventLoop(eventLoop, () -> {
+                for (RedisCommand<?, ?, ?> command : commands) {
+                    channelWrite(chan, command).addListener(AtMostOnceWriteListener.newInstance(this, command));
+                }
+                channelFlush(chan);
+            });
+        } else if (reliability == Reliability.AT_LEAST_ONCE) {
+            executeInEventLoop(eventLoop, () -> {
+                for (RedisCommand<?, ?, ?> command : commands) {
+                    channelWrite(chan, command).addListener(RetryListener.newInstance(this, command));
+                }
+                channelFlush(chan);
+            });
         }
+    }
 
-        if (reliability == Reliability.AT_LEAST_ONCE) {
-
-            // commands are ok to stay within the queue, reconnect will retrigger them
-            for (RedisCommand<?, ?, ?> command : commands) {
-                channelWrite(command).addListener(RetryListener.newInstance(this, command));
-            }
+    private void executeInEventLoop(EventLoop eventLoop, Runnable runnable) {
+        if (eventLoop.inEventLoop()) {
+            runnable.run();
+        } else {
+            eventLoop.execute(runnable);
```

Comment:
Looks like it will limit the throughput because it will has only few thread write, in the legacy way, both IOGroup and ExecutorGroup can execute.

can you verify the performance improvement more than JMH?